internal/db/schema/migrations/oss/postgres/28/01_prior_session_trigger.up.sql

```diff
@@ -0,0 +1,82 @@
+do $$
+declare
+    target record;
+begin
+    for target in select session_id from state_counts where count = 4 loop
+        -- If the last state is not terminated for this session...
+        if (select state from session_state where session_id in
+            (select session_id from state_counts where count = 4)
+            and session_id=target.session_id order by start_time desc limit 1)
+            != 'terminated' then
+            -- Prune invalid states after the session terminated
+            delete from session_state where session_id = target.session_id and previous_end_time >(
+                select previous_end_time from session_state where session_id=target.session_id and state='terminated');
+            -- Then remove terminated end time
+            update session_state set end_time=NULL where session_id=target.session_id and state='terminated';
+        end if;
+    end loop;
+end; $$;
+
+-- Remove bad states for session ids with 3 states, similar to the above
+-- Difference from above is the need to check if terminated exists in the set of states
+-- Ex: PTA, PTC -> pruned to PT, PT
+-- Valid transitions like PAT, PCT, and in progress sessions like PAC will be ignored
+do $$
+declare
+    target record;
+begin
+    for target in select session_id from state_counts where count = 3 loop
+        -- If the last state is not terminated for this session...
+        if (select state from session_state where session_id in
+            (select session_id from state_counts where count = 3)
+            and session_id=target.session_id order by start_time desc limit 1)
+            != 'terminated' then
+            -- See if terminated appears; if so, prune back to it
+            if exists(select * from session_state where session_id=target.session_id and state='terminated')then
+                --Then we find the terminated record and timestamp and delete those that came before
+                delete from session_state where session_id = target.session_id and previous_end_time >(
+                    select previous_end_time from session_state where session_id=target.session_id and state='terminated');
+                -- Then remove terminated end time
+                update session_state set end_time=NULL where session_id=target.session_id and state='terminated';
+            end if;
+        end if;
+    end loop;
+end; $$;
+
+-- Update session state transition trigger
+create function
+    update_prior_session_state()
+    returns trigger
+as $$
+begin
+    -- Prior state is the most recent state entry for this session_id
+    -- Invalid transitions will cause a constraint error on insert
+    new.prior_state = query.state from(
+         select state from session_state where session_id=new.session_id order by start_time desc limit 1) as query;
+
+    if new.prior_state is null then
+        new.prior_state='pending';
+    end if;
+
+    return new;
+end;
+$$ language plpgsql;
```

Comment:
Can we update the [`insert_session_state` trigger](https://github.com/hashicorp/boundary/blob/a8093b65e6c192f0517bdedc909583d80abff82f/internal/db/schema/migrations/oss/postgres/0/50_session.up.sql#L430-L454) to also handle the `prior_state` instead of creating a new trigger?